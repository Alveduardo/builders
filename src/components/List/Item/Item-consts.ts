export const ITEM = {
  TEST_ID: {
    CONTAINER: 'ITEM_CONTAINER_TEST_ID',
    LABEL: 'ITEM_LABEL_TEST_ID',
    ICON_BUTTON: 'ITEM_ICON_BUTTON_TEST_ID',
    TEMP_INDICATOR: 'ITEM_TEMP_INDICATOR_TEST_ID',
  },
};
