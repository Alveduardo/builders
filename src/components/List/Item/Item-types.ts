import { TextStyle, ViewStyle } from 'react-native';

interface DefaultStyles {
  containerStyle: ViewStyle;
  labelStyle: TextStyle;
  iconStyle: ViewStyle;
  tempIndicatorStyle: ViewStyle;
}

export interface ItemStyles {
  defaultStyles: DefaultStyles;
}

interface StateConfig {
  hour: string;
  iconName: string;
}

type StylesConfig = DefaultStyles;

export interface ItemConfig {
  state: StateConfig;
  styles: StylesConfig;
}

export interface ItemT {
  dt_txt: string;
  main: {
    temp: number;
    temp_min: number;
    temp_max: number;
  };
  weather: [
    {
      description: string;
    },
  ];
  rain?: {};
}

export interface ItemProps {
  item: ItemT;
  index: number;
}
