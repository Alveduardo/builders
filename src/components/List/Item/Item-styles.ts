import { ItemStyles } from './Item-types';

export const itemStyles: ItemStyles = {
  defaultStyles: {
    containerStyle: {
      paddingHorizontal: 32,
      width: '100%',
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      marginBottom: 16,
    },
    iconStyle: {
      flex: 1,
    },
  },
};
