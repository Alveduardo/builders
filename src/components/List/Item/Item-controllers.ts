import { getDateFromString, getPeriod } from '../../../contexts/Time/Time-utils';
import { isIncludedWord } from '../../../utils';
import { getIconWeather } from '../../../utils/weather/weather-utils';
import { itemStyles } from './Item-styles';
import { ItemConfig, ItemProps } from './Item-types';

export const useItemConfig = ({ item }: Pick<ItemProps, 'item'>): ItemConfig => {
  const {
    defaultStyles: { containerStyle, iconStyle },
  } = itemStyles;

  const split = item.dt_txt?.split(' ');
  const hour = split[1]?.substring(0, 5);

  const period = getPeriod(getDateFromString(item.dt_txt));
  const isRain = item?.rain ? true : false;

  const iconName = getIconWeather(period, isRain, isIncludedWord(item.weather[0].description, 'limpo'));

  return {
    state: {
      hour,
      iconName,
    },
    styles: {
      containerStyle,
      iconStyle,
    },
  };
};
