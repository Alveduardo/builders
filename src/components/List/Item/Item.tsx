import React, { memo } from 'react';
import { View } from 'react-native';

import Label from '../../Label';
import IconButton from '../../IconButton';
import TempIndicator from '../../TempIndicator';

import { Icon } from 'react-native-vector-icons/Icon';
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5';

import { ITEM } from './Item-consts';
import { ItemProps } from './Item-types';
import { useItemConfig } from './Item-controllers';

import { areEqualListProps } from '../../../utils';

const Item = memo(({ item, index }: ItemProps): JSX.Element => {
  const {
    state: { hour, iconName },
    styles: { containerStyle, iconStyle },
  } = useItemConfig({ item });

  return (
    <View testID={ITEM.TEST_ID.CONTAINER} style={containerStyle}>
      <Label testID={ITEM.TEST_ID.LABEL} size='xSmall'>
        {hour}
      </Label>
      <IconButton
        testID={ITEM.TEST_ID.ICON_BUTTON}
        type={FontAwesome5 as typeof Icon}
        name={iconName}
        size={'xxxSmall'}
        style={iconStyle}
        disabled
      />
      <TempIndicator
        testID={ITEM.TEST_ID.TEMP_INDICATOR}
        min={Math.round(item.main.temp_min)}
        max={Math.round(item.main.temp_max)}
      />
    </View>
  );
}, areEqualListProps);

export default Item;
