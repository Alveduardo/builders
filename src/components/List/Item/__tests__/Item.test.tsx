import React from 'react';
import { render } from '@testing-library/react-native';
import { ItemT } from '../Item-types';
import Item from '../Item';

const itemPropsDefault = {
  item: {
    dt_txt: '',
    main: {
      temp: 0,
      temp_min: 0,
      temp_max: 0,
    },
    weather: [
      {
        description: '',
      },
    ],
  } as ItemT,
  index: 0,
};

jest.mock('react-native-linear-gradient', () => 'LinearGradient');

describe('Item ->', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error').mockImplementation(jest.fn());
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.resetAllMocks();
  });

  describe('Snapshot ->', () => {
    it('Deve corresponder ao componente com as props default', () => {
      const tree = render(<Item {...itemPropsDefault} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('Deve corresponder ao componente com as props mais utilizadas', () => {
      const tree = render(<Item {...itemPropsDefault} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
