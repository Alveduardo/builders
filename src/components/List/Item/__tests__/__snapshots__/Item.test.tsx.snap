// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item -> Snapshot -> Deve corresponder ao componente com as props default 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "marginBottom": 16,
      "paddingHorizontal": 32,
      "width": "100%",
    }
  }
  testID="ITEM_CONTAINER_TEST_ID"
>
  <Text
    style={
      Array [
        Object {
          "color": "#FFF",
          "fontSize": 14,
        },
        undefined,
      ]
    }
    testID="ITEM_LABEL_TEST_ID"
  />
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
    testID="ITEM_ICON_BUTTON_TEST_ID"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#FFF",
            "fontSize": 16,
          },
          undefined,
          Object {
            "fontFamily": "FontAwesome5Free-Solid",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {
            "fontWeight": "900",
          },
        ]
      }
      testID="ICON_BUTTON_ICON_TEST_ID"
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
    testID="ITEM_TEMP_INDICATOR_TEST_ID"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFF",
            "fontSize": 16,
          },
          undefined,
        ]
      }
      testID="TEMP_INDICATOR_MIN_TEST_ID"
    >
      0
      °
    </Text>
    <LinearGradient
      colors={
        Array [
          "#b6f29b",
          "#ff6a38",
        ]
      }
      end={
        Object {
          "x": 1,
          "y": 0,
        }
      }
      start={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      style={
        Object {
          "borderRadius": 999,
          "height": 6,
          "marginHorizontal": 8,
          "width": 80,
        }
      }
      testID="TEMP_INDICATOR_GRADIENT_TEST_ID"
    />
    <Text
      style={
        Array [
          Object {
            "color": "#FFF",
            "fontSize": 16,
          },
          undefined,
        ]
      }
      testID="TEMP_INDICATOR_MAX_TEST_ID"
    >
      0
      °
    </Text>
  </View>
</View>
`;

exports[`Item -> Snapshot -> Deve corresponder ao componente com as props mais utilizadas 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "marginBottom": 16,
      "paddingHorizontal": 32,
      "width": "100%",
    }
  }
  testID="ITEM_CONTAINER_TEST_ID"
>
  <Text
    style={
      Array [
        Object {
          "color": "#FFF",
          "fontSize": 14,
        },
        undefined,
      ]
    }
    testID="ITEM_LABEL_TEST_ID"
  />
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
    testID="ITEM_ICON_BUTTON_TEST_ID"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#FFF",
            "fontSize": 16,
          },
          undefined,
          Object {
            "fontFamily": "FontAwesome5Free-Solid",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {
            "fontWeight": "900",
          },
        ]
      }
      testID="ICON_BUTTON_ICON_TEST_ID"
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
    testID="ITEM_TEMP_INDICATOR_TEST_ID"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFF",
            "fontSize": 16,
          },
          undefined,
        ]
      }
      testID="TEMP_INDICATOR_MIN_TEST_ID"
    >
      0
      °
    </Text>
    <LinearGradient
      colors={
        Array [
          "#b6f29b",
          "#ff6a38",
        ]
      }
      end={
        Object {
          "x": 1,
          "y": 0,
        }
      }
      start={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      style={
        Object {
          "borderRadius": 999,
          "height": 6,
          "marginHorizontal": 8,
          "width": 80,
        }
      }
      testID="TEMP_INDICATOR_GRADIENT_TEST_ID"
    />
    <Text
      style={
        Array [
          Object {
            "color": "#FFF",
            "fontSize": 16,
          },
          undefined,
        ]
      }
      testID="TEMP_INDICATOR_MAX_TEST_ID"
    >
      0
      °
    </Text>
  </View>
</View>
`;
